feat(ia): Newspack > Settings > Syndication Wizard

jaredrethman opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

* Migrated Syndication components to new Information Architecture.
   * Converted to TypeScript.
   * Refactored to use new components and hooks
* Added new hook `useWizardApiFetchToggle` for toggle (on/off) specific cards. See [`<Rss />`](https://github.com/Automattic/newspack-plugin/blob/feat/ia-settings-syndication/assets/wizards/newspack/views/settings/syndication/rss.tsx) component for usage.
* Refactored types to be more reusable i.e. `WizardTab`, `ApiMethods` and `PluginResponse`.
* Added `Configured` property to `plugins/configure` endpoint.
* Return `true` (for `is_configured` method) if no associated configuration manager class is found.

**Plugin Action Cards:**
* Refactored schema for `actionText` plugin cards. Ability to add select text based on plugin status.
* Added `isConfigurable` prop to allow 1 click configuration.
* Added two more callbacks `deactivate` & `configure`.
* `isTogglable` functionality to allow toggle (deactivate/activate) behavior for plugin cards.

### How to test the changes in this Pull Request:

1. Checkout this branch and build assets - `git checkout origin/feat/ia-settings-syndication && npm run build`
2. Navigate to _/wp-admin/admin.php?page=newspack-settings#/syndication_
3. Observe new tab "Syndication" loads correctly and it has 3 components; Rss, Apple News and Distributor (Appearance will vary depending on what plugin you have installed and activated.):

![Screenshot 2024-07-15 at 09 19 45](https://github.com/user-attachments/assets/0014f32b-982b-4d93-9234-f1dee2dadcfc)

4. Test:
   * Toggling on/off individual cards. Page will refresh after actions are completed (necessary for plugin related menu items to become visible/invisible).
   * Delete all related Apple News storage:
```sql
SELECT * FROM `wp_options` WHERE `option_name` LIKE '%apple_news%'
```
   * Refresh page and activate Apple News
![Screenshot 2024-07-15 at 09 33 51](https://github.com/user-attachments/assets/65d425ef-5754-4d85-991f-7bfd629c4837)
   * Run `wp option delete newspack_has_configured_plugin_publish-to-apple-news` in the terminal and refresh browser again. You should see an option to configure i.e. "Configure Apple News".
![Screenshot 2024-07-15 at 09 35 14](https://github.com/user-attachments/assets/cd0bc0e1-3d09-4527-89bd-0753f3b08c71)
   * Go ahead and click "Configure Apple News". 
   * After clicking - running `wp option get newspack_has_configured_plugin_publish-to-apple-news` should yield `1`, this is confirmation that Apple News has been configured.
![Screenshot 2024-07-15 at 09 40 02](https://github.com/user-attachments/assets/6fb45e02-c895-42fc-b5fd-5e8883177ccb)


### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

jaredrethman committed:
feat(ia): added plugin toggle card and various initial syndication pieces

jaredrethman committed:
Merge remote-tracking branch 'origin/epic/ia' into feat/ia-settings-syndication

jaredrethman committed:
Merge remote-tracking branch 'origin/epic/ia' into feat/ia-settings-syndication

jaredrethman committed:
feat(ia): add toggle to `WizardPluginToggleCard`

jaredrethman committed:
feat(ia): initial configurability conditional

jaredrethman committed:
Merge remote-tracking branch 'origin/epic/ia' into feat/ia-settings-syndication

jaredrethman committed:
refactor(ia): move toggle logic into plugin card

jaredrethman committed:
feat(ia): adding `useWizardApiToggle` hook for basic enable/disable functionality

jaredrethman committed:
refactor(ia): added reusable `ApiMethod` type

jaredrethman committed:
feat(ia): rss card support

jaredrethman committed:
refactor(ia): plugin support for new structure

jaredrethman committed:
refactor: configuration manager, return true

if no associated class is found

jaredrethman committed:
refactor(ia): prettier, comments

jaredrethman committed:
fix(ia): ci / prettier

dkoo reviewed (changes requested):
@jaredrethman I like the conceptual changes here, but it seems like not everything is working as intended:

- I can toggle ON the RSS Enhancement card, but once toggled on I'm unable to toggle it OFF—clicking the toggle has no effect.
- Clicking "Configure Apple News" does set the `newspack_has_configured_plugin_publish-to-apple-news` option to `1`, but there's no action in the UI—shouldn't it redirect to the `editLink` URL after clicking?
  - Related: This is currently true in `trunk` too, but I think the Apple News `editLink` should go to `admin.php?page=apple-news-options` (the plugin's settings page) instead of `admin.php?page=apple_news_index` (the plugin's index page, which just points you to the settings page if you haven't configured it yet)
- If toggling on a plugin card results in an error message, the page shouldn't do the refresh because then you can only see the error message for a split second.

jaredrethman committed:
fix(ia): pr feedback - rss allow toggle off

jaredrethman committed:
refactor(ia): pr feedback - disallow reload if error

jaredrethman committed:
refactor(ia): pr feedback - update apply editLink

jaredrethman committed:
feat(ia): pr feedback - redirect to editLink on configure

jaredrethman committed:
refactor(ia): pr feedback - disallow reload if error

jaredrethman commented:
> @jaredrethman I like the conceptual changes here, but it seems like not everything is working as intended:
> 
> * I can toggle ON the RSS Enhancement card, but once toggled on I'm unable to toggle it OFF—clicking the toggle has no effect.

Oh man, apologies for rookie mistake. Fixed!

> * Clicking "Configure Apple News" does set the `newspack_has_configured_plugin_publish-to-apple-news` option to `1`, but there's no action in the UI—shouldn't it redirect to the `editLink` URL after clicking?
>   
>   * Related: This is currently true in `trunk` too, but I think the Apple News `editLink` should go to `admin.php?page=apple-news-options` (the plugin's settings page) instead of `admin.php?page=apple_news_index` (the plugin's index page, which just points you to the settings page if you haven't configured it yet)

Updated. Upon clicking "Configure X" page will redirect to associated `editLink`

> * If toggling on a plugin card results in an error message, the page shouldn't do the refresh because then you can only see the error message for a split second.

Nice catch, have amended related hook and component.

dkoo reviewed (approved):
Thanks for addressing those issues, @jaredrethman! This is working nicely now.
